File: 1372-longest-zigzag-path-in-a-binary-tree/1372-longest-zigzag-path-in-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    void zigzagPath(TreeNode *root, int leftPath, int rightPath, int &ans) {
        if(root == NULL) return;
        ans = max(ans, max(leftPath, rightPath));
        zigzagPath(root->left, 0, leftPath + 1, ans);
        zigzagPath(root->right, rightPath + 1, 0, ans);
    }
public:
    int longestZigZag(TreeNode* root) {
        int leftPath = 0, rightPath = 0, ans = 0;
        zigzagPath(root, leftPath, rightPath, ans);
        return ans;
    }
};

File: 1372-longest-zigzag-path-in-a-binary-tree/1372-longest-zigzag-path-in-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def longestZigZag(self, root: Optional[TreeNode]) -> int:
        l, r, ans = 0, 0, [0]
        def dfs(root, l, r, ans):
            if root == None: return
            ans[0] = max(ans[0], l, r)
            dfs(root.left, 0, l + 1, ans)
            dfs(root.right, r + 1, 0, ans)
        dfs(root, l, r, ans)
        return ans[0]